tools/tinygoize/main.go

```diff
@@ -2,81 +2,340 @@
+// despite being in different directories.
+var addBuildTags = map[string]string{
+	"gzip":     "noasm",
+	"insmod":   "noasm",
+	"rmmod":    "noasm",
+	"bzimage":  "noasm",
+	"kconf":    "noasm",
+	"modprobe": "noasm",
+	"console":  "noasm",
+	"init":     "noasm",
+}
+
+// returns the needed build-tags for a given package
+func buildTags(dir string) (tags string) {
+	parts := strings.Split(dir, "/")
+	cmd := parts[len(parts)-1]
+	return addBuildTags[cmd]
+}
+
+// Track set of passing, failing, and excluded commands
+type BuildStatus struct {
+	passing  []string
+	failing  []string
+	excluded []string
+}
+
+// return trimmed output of "tinygo version"
+func tinygoVersion(tinygo *string) (string, error) {
+	out, err := exec.Command(*tinygo, "version").CombinedOutput()
+	if nil != err {
+		return "", err
+	}
+	return strings.TrimSpace(string(out)), err
+}
+
+// check (via `go build -n`) if a given directory would have been skipped
+// due to build constraints (e.g. cmds/core/bind only builds for plan9)
+func isExcluded(dir string) bool {
+	// too lazy to dynamically pull tags from `tinygo info`
+	tags := []string{
+		"tinygo",
+		"tinygo.enable",
+		"purego",
+		"osusergo",
+		"math_big_pure_go",
+		"gc.precise",
+		"scheduler.tasks",
+		"serial.none",
+	}
+	c := exec.Command("go", "build",
+		"-n",
+		"-tags", strings.Join(tags, ","),
+	)
+	c.Env = append(os.Environ(), "GOOS=linux", "GOARCH=amd64")
+	c.Dir = dir
+	out, _ := c.CombinedOutput()
+	return strings.Contains(string(out), "build constraints exclude all Go files in")
+}
+
+// "tinygo build" in directory 'dir'
+func build(tinygo *string, dir string) (err error) {
+	tags := []string{"tinygo.enable"}
+	if addTags := buildTags(dir); addTags != "" {
+		tags = append(tags, addTags)
+	}
+	c := exec.Command(*tinygo, "build", "-tags", strings.Join(tags, ","))
+	c.Dir = dir
+	c.Stdout, c.Stderr = os.Stdout, os.Stderr
+	c.Env = append(os.Environ(), "GOOS=linux", "CGO_ENABLED=0", "GOARCH=amd64")
+	return c.Run()
+}
+
+// "tinygo build" in each of directories 'dirs'
+func buildDirs(tinygo *string, dirs []string) (status BuildStatus, err error) {
+	for _, dir := range dirs {
+		log.Printf("%s Building...\n", dir)
+		err = build(tinygo, dir)
+		if err == nil {
+			log.Printf("%v PASS\n", dir)
+			status.passing = append(status.passing, dir)
+		} else {
+			berr, ok := err.(*exec.ExitError)
+			if !ok {
+				break
+			}
+			err = nil
+			if isExcluded(dir) {
+				log.Printf("%v EXCLUDED\n", dir)
+				status.excluded = append(status.excluded, dir)
+			} else {
+				log.Printf("%v FAILED %v\n", dir, berr)
+				status.failing = append(status.failing, dir)
+			}
+		}
+	}
+	return
+}
 
+// Modifies, adds, or removes //go:build line as appropriate to include / remove
+// '!tinygo || tinygo.enable' for all .go files in dir depending on whether it
+// 'builds' or not as previously tested.
+func fixupConstraints(dir string, builds bool) (err error) {
```

Comment:
The functionality this function provides adds a lot of complex functionality. Especially since it works on the AST of the go programs and does literal modifications and checks a lot of edge cases we should add some basic test cases to ensure the modifications are done correctly. Since this PR has been in review for quite a while and the except my suggestions look good to me, we could also bring in the tests in a follow-up PR.

Reply:
Yeah, might get to some units on the AST checks.  Shouldn't run this thing on anything but a clean tree and shouldn't commit anything it modifies without visual inspection. In that vein, not sure of the value of units. Part of me finds this an overly complicated replacement for a perl one-liner.